### PR TITLE
solved some npu bugs

### PR DESCRIPTION
**Pull request by Baibaifan**
### PR types

Bug fixes
### PR changes

Others
### Describe
 
shared_index api support two types int32 and int64. 
Input indices with data type int64 and int32. It's last dimension must be 1.
situation1:
```
 import paddle
            label = paddle.to_tensor([[16], [1]], "int64")
            shard_label = paddle.shard_index(input=label,
                                             index_num=20,
                                             nshards=2,
                                             shard_id=0)
            print(shard_label)
```
situation2:
```
 import paddle
            label = paddle.to_tensor([[16], [1]], "int32")
            shard_label = paddle.shard_index(input=label,
                                             index_num=20,
                                             nshards=2,
                                             shard_id=0)
            print(shard_label)
````

solved some npu bugs

### CODE REVIEW

File: python/paddle/distributed/fleet/meta_optimizers/sharding/utils.py [hunk 1]
```diff
@@ -402,13 +402,16 @@ def get_grad_device(grad_name, shard):
...
-def get_first_check_finite_and_unscale_op_idx(block):
+def get_first_check_finite_and_unscale_op_idx(block, raise_error=True):
 
     for idx, op in enumerate(block.ops):
         if op.type == "check_finite_and_unscale":
             return idx
 
-    raise ValueError("check_finite_and_unscale does not exist in block")
+    if raise_error:
+        raise ValueError("check_finite_and_unscale does not exist in block")
```
Comment (JZ-LIANG) [inline python/paddle/distributed/fleet/meta_optimizers/sharding/utils.py]: the error message should be:
"amp is turn on but check_finite_and_unscale op does not exist in main block"

File: python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py [hunk 2]
```diff
@@ -309,14 +309,15 @@ def minimize_impl(self,
...
                 first_optimize_op_index = get_first_check_finite_and_unscale_op_idx(
-                    main_block)
-                insert_allreduce_ops(
-                    main_block,
-                    first_optimize_op_index,
-                    self.dp_ring_id,
-                    accumulated_grad_names,
-                    core.op_proto_and_checker_maker.OpRole.Optimize,
-                    use_calc_stream=True)
+                    main_block, raise_error=self.user_defined_strategy.amp)
```
Comment (JZ-LIANG) [inline python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py:312]: what is the reason for this modification ? for npu hang bug?

**Review by JZ-LIANG — CHANGES_REQUESTED** [review]

File: python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py (hunk 2 above)
Comment (Baibaifan) [reply python/paddle/distributed/fleet/meta_optimizers/sharding_optimizer.py:312]: To solove nup hang bugs!

File: python/paddle/distributed/fleet/meta_optimizers/sharding/utils.py (hunk 1 above)
Comment (Baibaifan) [reply python/paddle/distributed/fleet/meta_optimizers/sharding/utils.py]: ok

File: paddle/fluid/operators/collective/recv_v2_op_npu.cc [hunk 3]
```diff
@@ -28,6 +28,7 @@ class CRecvOpASCENDKernel : public framework::OpKernel<T> {
   void Compute(const framework::ExecutionContext& ctx) const override {
 #if defined(PADDLE_WITH_ASCEND_CL)
     auto x = ctx.Output<framework::LoDTensor>("Out");
```
Comment (zhiqiu) [inline paddle/fluid/operators/collective/recv_v2_op_npu.cc]:
```suggestion
    auto out = ctx.Output<framework::LoDTensor>("Out");
```

File: python/paddle/nn/functional/common.py [hunk 4]
```diff
@@ -1467,7 +1467,7 @@ def linear(x, weight, bias=None, name=None):
         }
         tmp = helper.create_variable_for_type_inference(dtype)
         helper.append_op(
-            type='matmul', inputs=inputs, outputs={'Out': tmp}, attrs=attrs)
+            type='matmul_v2', inputs=inputs, outputs={'Out': tmp}, attrs=attrs)
```
Comment (zhiqiu) [inline python/paddle/nn/functional/common.py]: Why change here? Is it consistent?

File: python/paddle/distributed/fleet/meta_optimizers/sharding/utils.py [hunk 5]
```diff
@@ -402,13 +402,18 @@ def get_grad_device(grad_name, shard):
...
+def get_first_check_finite_and_unscale_op_idx(block, raise_error=True):
 
     for idx, op in enumerate(block.ops):
         if op.type == "check_finite_and_unscale":
             return idx
 
-    raise ValueError("check_finite_and_unscale does not exist in block")
+    if raise_error:
+        raise ValueError(
+            "amp is turn on but check_finite_and_unscale op does not exist in main block"
```
Comment (zhiqiu) [inline python/paddle/distributed/fleet/meta_optimizers/sharding/utils.py]: turn -> turned

**Review by JZ-LIANG — DISMISSED** [review]
LGTM for Sharding

File: paddle/fluid/framework/operator.cc [hunk 6]
```diff
@@ -1299,7 +1301,9 @@ void OperatorWithKernel::TransferInplaceVarsBack(
     auto* transformed_tensor = GetLoDTensorOrSelectedRowsValueFromVar(*var);
     auto original_dims = original_tensor->dims();
     original_tensor->ShareDataWith(*transformed_tensor);
-    original_tensor->Resize(original_dims);
+    if (type_ != "reshape2" && type_ != "reshape2_grad") {
+      original_tensor->Resize(original_dims);
+    }
```
Comment (zhiqiu) [inline paddle/fluid/framework/operator.cc:1308]: As discussed, it is a temp solution to change here, please add some notes.

File: python/paddle/fluid/dataset.py [hunk 7]
```diff
@@ -251,6 +251,8 @@ def set_use_var(self, var_list):
                 slot_var.type = "float"
             elif var.dtype == core.VarDesc.VarType.INT64:
                 slot_var.type = "uint64"
+            elif var.dtype == core.VarDesc.VarType.INT32:
+                slot_var.type = "uint32"
             else:
                 raise ValueError(
                     "Currently, fluid.dataset only supports dtype=float32 and dtype=int64"
```
Comment (sandyhouse) [inline python/paddle/fluid/dataset.py]: Add INT32?

File: python/paddle/fluid/optimizer.py [hunk 8]
```diff
@@ -4272,8 +4275,9 @@ def _check_validation(self, block):
                             "{} has not been set.".format(op.type))
             if device == "gpu:all": continue
             dev_type = device.split(':')[0]
-            assert dev_type == "gpu", ("Now only gpu devices are supported "
-                                       "for pipeline parallelism.")
+            assert dev_type == "gpu" or dev_type == 'npu', (
+                "Now only gpu and npu devices are supported "
+                "for pipeline parallelism.")
```
Comment (sandyhouse) [inline python/paddle/fluid/optimizer.py:4280]: How to deal with npu:all?

File: paddle/fluid/operators/collective/recv_v2_op_npu.cc [hunk 9]
```diff
@@ -27,10 +27,11 @@ class CRecvOpASCENDKernel : public framework::OpKernel<T> {
  public:
   void Compute(const framework::ExecutionContext& ctx) const override {
 #if defined(PADDLE_WITH_ASCEND_CL)
-    auto x = ctx.Output<framework::LoDTensor>("Out");
-    void* ptr = reinterpret_cast<void*>(const_cast<T*>(x->data<T>()));
-    int numel = x->numel();
-    HcclDataType dtype = platform::ToHCCLDataType(x->type());
+    auto out = ctx.Output<framework::LoDTensor>("Out");
+    out->mutable_data<T>(out->dims(), ctx.GetPlace());
+    void* ptr = reinterpret_cast<void*>(const_cast<T*>(out->data<T>()));
```
Comment (zhiqiu) [inline paddle/fluid/operators/collective/recv_v2_op_npu.cc:32]: Not important.
```suggestion
    void* ptr = out->data<void>();
```

**Review by zhiqiu — DISMISSED** [review]

**Review by zhiqiu — APPROVED** [review]
LGTM

**Review by jzhang533 — COMMENTED** [review]
我看shard_index有一个deprecate的说明。
所以对这个API的维护，是在fluid下面原地修改，还是需要迁移一下？

**Review by jzhang533 — APPROVED** [review]
lgtm

**Review by sandyhouse — APPROVED** [review]
LGTM for pp

**Pull request merged**